### DISCUSSION TITLE
JK-B2A24S20P HW 10.XG SW 10.07D30

### DISCUSSION THREAD

**Discussion by TheSmartGerman** [syssi/esphome-jk-bms#122] [Show and tell]
Well, 

just established a connection ESP32BLE and like to share the log (just deleted some personal macs/pw). I did some research and can't find to many information about HW 10

Also did the retrieve device and settings info.

I do see some unkonwn messages. 

I may think [15:44:22][I][jk_bms_ble:841]:   Unknown138: 3.450000 by be the balancer trigger voltage.

[logs_esp32-gateway-og_logs.txt](https://github.com/syssi/esphome-jk-bms/files/8832796/logs_esp32-gateway-og_logs.txt)

**Comment by syssi** [comment]
Could you change the balancer trigger voltage and check the value a second time to verify your guess?

**Reply by TheSmartGerman** [reply]
I changes, Change "Start Balance Volt. (V)", I sett 3,44V now and Unknown138 changes according to the value.

> [08:53:48][D][button:013]: 'esp32-gateway-og retrieve settings' Pressed.
> [08:53:48][D][jk_bms_ble:1075]: Write register: AA.55.90.EB.96.00.00.00.00.00.00.00.00.00.00.00.00.00.00.10 (20)
> [08:53:48][I][jk_bms_ble:746]: Settings frame (300 bytes):
> [08:53:48][I][jk_bms_ble:772]:   Unknown6: 0.600000
> [08:53:48][I][jk_bms_ble:774]:   Cell UVP: 2.600000 V
> [08:53:48][I][jk_bms_ble:776]:   Cell UVPR: 2.650000 V
> [08:53:48][I][jk_bms_ble:778]:   Cell OVP: 3.600000 V
> [08:53:48][I][jk_bms_ble:780]:   Cell OVPR: 3.550000 V
> [08:53:48][I][jk_bms_ble:782]:   Balance trigger voltage: 0.010000 V
> [08:53:48][I][jk_bms_ble:788]:   Power off voltage: 2.500000 V
> [08:53:48][I][jk_bms_ble:790]:   Max. charge current: 80.000000 A
> [08:53:48][I][jk_bms_ble:792]:   Charge OCP delay: 30.000000 s
> [08:53:48][I][jk_bms_ble:794]:   Charge OCP recovery delay: 60.000000 s
> [08:53:48][I][jk_bms_ble:796]:   Max. discharge current: 120.000008 A
> [08:53:48][I][jk_bms_ble:798]:   Discharge OCP recovery delay: 300.000000 s
> [08:53:48][I][jk_bms_ble:800]:   Discharge OCP recovery delay: 60.000000 s
> [08:53:48][I][jk_bms_ble:802]:   SCP recovery time: 60.000000 s
> [08:53:48][I][jk_bms_ble:804]:   Max. balance current: 2.000000 A
> [08:53:48][I][jk_bms_ble:806]:   Charge OTP: 70.000000 °C
> [08:53:48][I][jk_bms_ble:808]:   Charge OTP recovery: 60.000000 °C
> [08:53:49][I][jk_bms_ble:810]:   Discharge OTP: 70.000000 °C
> [08:53:49][I][jk_bms_ble:812]:   Discharge OTP recovery: 60.000000 °C
> [08:53:49][I][jk_bms_ble:814]:   Charge UTP: -20.000000 °C
> [08:53:49][I][jk_bms_ble:816]:   Charge UTP recovery: -10.000000 °C
> [08:53:49][I][jk_bms_ble:818]:   MOS OTP: 90.000000 °C
> [08:53:49][I][jk_bms_ble:820]:   MOS OTP recovery: 70.000000 °C
> [08:53:49][I][jk_bms_ble:822]:   Cell count: 16.000000
> [08:53:49][I][jk_bms_ble:825]:   Charge switch: on
> [08:53:49][I][jk_bms_ble:829]:   Discharge switch: on
> [08:53:49][I][jk_bms_ble:833]:   Balancer switch: on
> [08:53:49][I][jk_bms_ble:837]:   Nominal battery capacity: 280.000000 Ah
> [08:53:49][I][jk_bms_ble:839]:   Unknown134: 1.500000
> [08:53:49][I][jk_bms_ble:841]:   Unknown138: 3.440000
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 1: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 2: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 3: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 4: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 5: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 6: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 7: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 8: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 9: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 10: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 11: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 12: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 13: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 14: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 15: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 16: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 17: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 18: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 19: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 20: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 21: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 22: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 23: 0.000000 Ohm
> [08:53:49][I][jk_bms_ble:871]:   Con. wire resistance 24: 0.000000 Ohm

**Reply by syssi** [reply]
Thanks for your support! I've pushed a proper label now: https://github.com/syssi/esphome-jk-bms/commit/e9ed38b0a9fbc8235a99da9b18c22d75f3307d9f

**Reply by syssi** [reply]
Feel free to create an issue if you miss something or if you see some room for improvements.

**Comment by TheSmartGerman** [comment]
OK great!

Well, so far I just started to use your implementation and so far it works like a charm. 

Just some questions:
- TTL (RS485) and BLE, both JK02, protocol doesn't deliver the same data (BLE is more powerfull?)
- You thinking about to grap the TTL data from the bluetooth modul, so get a wired connection with the same datasets?
- Does the ESP Configuration (YAML) provied a possibility only send the values of the detected cells? (Sure I know I have 16 Cells and just can delete the voltage and resistance values)

Next will try your pisolar implementation ;)

**Reply by syssi** [reply]
1. There is a misunderstanding: TTL is another protocol (!= JK02) with different payload.
2. I want to grab the TTL data from the BLE (TX, RX) and turn off the BLE module by pulling RST to GND because the JK02 protocol is read-write. The TTL protocol is read-only.
3. At the JK02 there are two bytes indicating something like the cell count but the encoding is not known: https://github.com/syssi/esphome-jk-bms/blob/main/components/jk_bms_ble/jk_bms_ble.cpp#L405-L407

At the TTL protocol the cell count can be calculated: https://github.com/syssi/esphome-jk-bms/blob/main/components/jk_bms/jk_bms.cpp#L74

I don't want to register sensors automatically depending on the cell count because it removes a degree of freedom. You should just remove useless sensors from your configuration.
